src/metabase/driver/sql_jdbc/sync/describe_table.clj

```diff
@@ -402,20 +405,33 @@
           json-fields           (filter #(isa? (:base-type %) :type/JSON) table-fields)]
       (if (nil? (seq json-fields))
         #{}
-        (binding [hx/*honey-sql-version* (sql.qp/honey-sql-version driver)]
-          (let [json-field-names (mapv #(apply hx/identifier :field (into table-identifier-info [(:name %)])) json-fields)
-                table-identifier (apply hx/identifier :table table-identifier-info)
-                sql-args         (sql.qp/format-honeysql driver {:select (mapv sql.qp/maybe-wrap-unaliased-expr json-field-names)
-                                                                 :from   [(sql.qp/maybe-wrap-unaliased-expr table-identifier)]
-                                                                 :limit  metadata-queries/nested-field-sample-limit})
-                query            (jdbc/reducible-query spec sql-args {:identifiers identity})
-                field-types      (transduce describe-json-xform describe-json-rf query)
-                fields           (field-types->fields field-types)]
-            (if (> (count fields) max-nested-field-columns)
-              (do
-                (log/warn
-                 (format
-                  "More nested field columns detected than maximum. Limiting the number of nested field columns to %d."
-                  max-nested-field-columns))
-                (set (take max-nested-field-columns fields)))
-              fields)))))))
+        (let [existing-fields-by-name (m/index-by :name (t2/select 'Field :table_id (u/the-id table)))
+              unfold-json-fields      (filter (fn [field]
+                                                ;; unfold json if the field has json_unfolding = true
+                                                ;; don't unfold json if the field has json_unfolding = false
+                                                ;; otherwise if the field doesn't exist or json_unfolding = nil
+                                                ;; use the database default
+                                                (let [existing-field (existing-fields-by-name (:name field))]
+                                                  (or (:json_unfolding existing-field)
+                                                      (and (nil? (:json_unfolding existing-field))
```

Comment:
if we make `json_unfodling` non-nullable, we don't have to deal with it here.

Reply:
then this predicate will be only `(:json_unfolding existing-field)`

Reply:
:json_unfolding can still be nil if the existing field doesn't exist. I've updated it here to reflect that. https://github.com/metabase/metabase/commit/4c53a5b5d92055d3a77d35b9012bfa03bf361c28

Reply:
ah nice. that sound like a gotcha tho, should we have tests for it?

Reply:
It's covered by the `json-unfolding-default-false-test` test for json unfolding, since this code path is hit twice in that test. Once in the first sync when the json_unfolding default is false for the database, and again in the second sync when the json_unfolding default is true.